backend/courses/management/commands/recompute_parent_courses.py

```diff
@@ -0,0 +1,88 @@
+from textwrap import dedent
+
+from botocore.exceptions import NoCredentialsError
+from django.core.management.base import BaseCommand
+from django.db import transaction
+from django.db.models import F, Max, OuterRef, Subquery, Value
+
+from courses.management.commands.load_crosswalk import load_crosswalk
+from courses.models import Course
+from courses.util import get_semesters, in_dev
+
+
+def recompute_parent_courses(semesters: list[str], verbose=False):
+    """
+    Attemps to recompute empty `parent_course` fields for all courses in the specified semesters.
+
+    :param semesters: Semesters for which you want to recompute `parent_course` fields.
+    :param verbose: Whether to print status/progress updates.
+    """
+    if verbose:
+        print(f"Attempting to recompute empty `parent_course` fields for semesters {semesters}.")
+    filled_links = 0
+    for i, semester in enumerate(sorted(semesters)):
+        if verbose:
+            print(f"Processing semester {semester} ({i+1}/{len(semesters)})...")
+        with transaction.atomic():
+            courses = list(Course.objects.filter(semester=semester))
+            full_codes = {course.full_code for course in courses}
+            parent_candidates = (
+                Course.objects.filter(full_code__in=full_codes, semester__lt=semester)
+                .annotate(
+                    max_semester=Subquery(
+                        Course.objects.filter(
+                            full_code=OuterRef("full_code"), semester__lt=semester
+                        )
+                        .annotate(common=Value(1))
+                        .values("common")
+                        .annotate(max_semester=Max("semester"))
+                        .values("max_semester")
+                    )
+                )
+                .filter(semester=F("max_semester"))
+            )
+            parent_candidates = {p.full_code: p for p in parent_candidates}
+            for course in courses:
+                course.parent_course = parent_candidates.get(course.full_code)
+            Course.objects.bulk_update(courses, ["parent_course"])
+            if verbose:
+                print(f"Recomputed {len(courses)} `parent_course` links.")
+            filled_links += len(courses)
+    if verbose:
+        print(
+            f"Finished recomputing {filled_links} `parent_course` fields for semesters {semesters}."
+        )
+
+    try:
+        load_crosswalk(print_missing=False, verbose=verbose)
```

Comment:
Hmm yes BUT I don't think we can use the crosswalk to delete existing parent links because the crosswalk doesn't claim to have a comprehensive mapping of parent to child courses...

I believe the soft state topic script should handle multiple children fairly well. We could theoretically have some boolean flag denoting whether a link came from the crosswalk / manual input vs our automatic heuristic (rn just same full code)... that might be a good way to prevent this automatic script from overwriting manually confirmed links.